docs/proposals/202005_query_logging.md

```diff
@@ -0,0 +1,242 @@
+
+Another requirement is using **Audit and Adaptive Logging** in Thanos. This is a useful feature, as the former helps in aggregating all the queries made until now, and that can be utilised to track down and provide a comprehensive overview of what is happening. Audit logs can help in capturing the flow of requests made, thus we can track line to line flow of queries which can be helpful in describing the observed behaviour of a request. More info - https://rollout.io/blog/audit-logs/
+
+```log
+    {
+      protoPayload: {
+        @type: "type.googleapis.com/google.cloud.audit.AuditLog",
+        status: {},
+        authenticationInfo: {
+          principalEmail: "user@example.com"
+        },
+        serviceName: "appengine.googleapis.com",
+        methodName: "SetIamPolicy",
+        authorizationInfo: [...],
+        serviceData: {
+          @type: "type.googleapis.com/google.appengine.legacy.AuditData",
+          policyDelta: { bindingDeltas: [
+              action: "ADD",
+              role: "roles/logging.privateLogViewer",
+              member: "user:user@example.com"
+          ], }
+        },
+        request: {
+          resource: "my-gcp-project-id",
+          policy: { bindings: [...], }
+        },
+        response: {
+          bindings: [
+            {
+              role: "roles/logging.privateLogViewer",
+              members: [ "user:user@example.com" ]
+            }
+          ],
+        }
+      },
+      insertId: "53179D9A9B559.AD6ACC7.B40604EF",
+      resource: {
+        type: "gcp_app",
+        labels: { project_id: "my-gcp-project-id" }
+      },
+      timestamp: "2019-05-27T16:24:56.135Z",
+      severity: "NOTICE",
+      logName: "projects/my-gcp-project-id/logs/cloudaudit.googleapis.com%2Factivity",
+    }
+Taken from - https://cloud.google.com/logging/docs/audit/understanding-audit-logs
+```
+_Example of an audit logging_
+
+**Adaptive Logging** is quite useful, as we can log all queries based upon fulfilling a certain criteria/filters, and later those can be used for inspecting abnormal queries and improving the user experience overall.
+
+## Proposal
+
+### Query Logging in Thanos
+
+The proposal for Query Logging has been divided into three major parts - Audit Logging, Adaptive Logging and Active Query Logging. Audit and Adaptive Logging would be implemented across the Store API level, while the Active Query Logging would be local to each component of Thanos, providing logs of queries currently running.
+
+![](../img/thanos_log_limit.png)
+
+Each of the use case of the sub-part would be described and a possible implementation would be discussed -
+
```

Comment:
I think it's worth to mention specific agreements we have on what user can configure in Thanos:

* What adaptive policies
* What logging levels for requests
* How to turn off audit of request/adaptive
* How to redirect audit or adaptive to separate files

Reply:
* How to control what methods (if hardcoded  which methods are hardcoded)

Reply:
* how to control if you want start log entry or not

Reply:
This is resolved!